git/githistory/log/log.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	l := &Logger{
+		sink: sink,
+		widthFn: func() int {
+			size, err := ts.GetSize()
+			if err != nil {
+				return 80
+			}
+			return size.Col()
+		},
+		queue: make(chan Task),
+		tasks: make(chan Task),
+		wg:    new(sync.WaitGroup),
+	}
+
+	go l.consume()
+
+	return l
+}
+
+// Close closes the queue and does not allow new Tasks to be `enqueue()`'d. It
+// waits until the currently running Task has completed.
+func (l *Logger) Close() {
+	if l == nil {
+		return
+	}
+
+	close(l.queue)
+
+	l.wg.Wait()
+}
+
+// Waitier creates and enqueues a new *WaitingTask.
+func (l *Logger) Waiter(msg string) *WaitingTask {
+	t := NewWaitingTask(msg)
+	l.enqueue(t)
+
+	return t
+}
+
+// Percentage creates and enqueues a new *PercentageTask.
+func (l *Logger) Percentage(msg string, total uint64) *PercentageTask {
+	t := NewPercentageTask(msg, total)
+	l.enqueue(t)
+
+	return t
+}
+
+// enqueue enqueues the given Tasks "ts".
+func (l *Logger) enqueue(ts ...Task) {
+	if l == nil {
+		for _, t := range ts {
+			go func(t Task) {
+				for range <-t.Updates() {
+					// Discard all updates.
+				}
+			}(t)
+		}
+		return
+	}
+
+	l.wg.Add(len(ts))
+	for _, t := range ts {
+		l.queue <- t
+	}
+}
+
+// consume creates a pseudo-infinte buffer between the incoming set of tasks and
+// the queue of tasks to work on.
+func (l *Logger) consume() {
+	go func() {
+		// Process the single next task in sequence until completion,
+		// then consume the next task.
+		for task := range l.tasks {
+			l.logTask(task)
+		}
+	}()
+
+	pending := make([]Task, 0)
+
+L:
+	for {
+		// If there is a pending task, "peek" it off of the set of
+		// pending tasks.
+		var next Task
+		if len(pending) > 0 {
+			next = pending[0]
+		}
+
+		if next == nil {
+			// If there was no pending task, wait for either a)
+			// l.queue to close, or b) a new task to be submitted.
+			task, ok := <-l.queue
+			if !ok {
+				// If the queue is closed, no more new tasks may
+				// be added.
+				break L
+			}
+
+			// Otherwise, add a new task to the set of tasks to
+			// process immediately, since there is no current
+			// buffer.
+			l.tasks <- task
+		} else {
+			// If there is a pending task, wait for either a) a
+			// write to process the task to become non-blocking, or
+			// b) a new task to enter the queue.
+			select {
+			case task, ok := <-l.queue:
+				if !ok {
+					// If the queue is closed, no more tasks
+					// may be added.
+					break L
+				}
+				// Otherwise, add the next task to the set of
+				// pending, active tasks.
+				pending = append(pending, task)
+			case l.tasks <- next:
+				// Or "pop" the peeked task off of the pending
+				// set.
+				pending = pending[1:]
+			}
+		}
+	}
+
+	close(l.tasks)
+}
+
+// logTask logs the set of updates from a given task to the sink, then logs a
+// "done" message, and then marks the task as done.
+func (l *Logger) logTask(task Task) {
+	defer l.wg.Done()
+
+	var last string
+	for last = range task.Updates() {
+		l.logLine(last)
```

Comment:
It may be worth watching your CPU on an intense write load. The existing spinner that `fetch` updates the terminal no more frequently than every 200ms. It's plenty time to represent progress to humans, and doesn't write to the terminal nearly as frequently.